### PR TITLE
Assignment8

### PR DESCRIPTION
**Pull request by Cybotive**

### CODE REVIEW

File: Assignment/SampleData.cs [hunk 1]
```diff
@@ -1,16 +1,27 @@
...
 
 namespace Assignment
 {
     public class SampleData : ISampleData
     {
         // 1.
-        public IEnumerable<string> CsvRows => throw new NotImplementedException();
+        public IEnumerable<string> CsvRows
+            => from line in File.ReadAllLines("People.csv")
+               where !line.StartsWith("Id")
```
Comment (dsergio) [inline Assignment/SampleData.cs]: clever

File: Assignment/SampleData.cs [hunk 2]
```diff
@@ -1,16 +1,27 @@
...
+               select line;
 
         // 2.
-        public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() 
-            => throw new NotImplementedException();
+        public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()
+        {
+            IEnumerable<string> query = (from row in CsvRows
+                                        let state = row.Split(',')[6]
```
Comment (dsergio) [inline Assignment/SampleData.cs]: you can use an enum Column { Id, Address, etc ... }

File: Assignment/SampleData.cs [hunk 3]
```diff
@@ -1,16 +1,27 @@
...
+                                        select state).Distinct().OrderBy(item => item);
+
+            return query;
+        }
```
Comment (dsergio) [inline Assignment/SampleData.cs]: don't forget the rest of the implementation

File: Assignment.Tests/SampleDataTest.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            foreach (string s in states)
+            {
+                int count = 0;
+                foreach (string b in states)
+                {
+                    if (b == s)
+                        count++;
+                }
+
+                Assert.AreEqual(1, count);
```
Comment (dsergio) [inline Assignment.Tests/SampleDataTest.cs]: nice

File: Assignment/SampleData.cs [hunk 5]
```diff
@@ -1,30 +1,86 @@
...
-            Predicate<string> filter) => throw new NotImplementedException();
+            Predicate<string> filter)
+        {
+            return from person in People
+                   where filter(person.EmailAddress)
+                   select (person.FirstName, person.LastName);
+        }
 
         // 6.
         public string GetAggregateListOfStatesGivenPeopleCollection(
```
Comment (MarkMichaelis) [inline Assignment/SampleData.cs]: Member can be made static

File: Assignment/SampleData.cs [hunk 6]
```diff
@@ -1,30 +1,86 @@
...
 using System.Collections.Generic;
+using System.IO;
+using System.Linq;
 
 namespace Assignment
 {
     public class SampleData : ISampleData
     {
         // 1.
-        public IEnumerable<string> CsvRows => throw new NotImplementedException();
```
Comment (MarkMichaelis) [inline Assignment/SampleData.cs]: Member can be made static

File: Assignment/SampleData.cs [hunk 7]
```diff
@@ -1,30 +1,86 @@
...
+                List<Person> people = new List<Person>();
+                IEnumerable<string> rows = CsvRows;
+
+                foreach (string s in rows)
+                {
+                    Person pers = new Person();
+                    Address addr = new Address();
+                    string[] items = s.Split(',');
+
+                    pers.FirstName = items[1];
```
Comment (MarkMichaelis) [inline Assignment/SampleData.cs]: Avoid hard coded values (magic values) like an integer for the column. They are hard to read and interpret.

File: Assignment/SampleData.cs [hunk 8]
```diff
@@ -1,30 +1,86 @@
...
+                                        select state).Distinct().OrderBy(item => item);
+
+            return query;
+        }
 
         // 3.
         public string GetAggregateSortedListOfStatesUsingCsvRows()
-            => throw new NotImplementedException();
+        {
+            string[] stateArr = GetUniqueSortedListOfStatesGivenCsvRows().ToArray();
```
Comment (MarkMichaelis) [inline Assignment/SampleData.cs]: You can use IEnumerabe<string>, not need to call ToArray().

**Comment by MarkMichaelis** [comment]
##  (Cybotive)

## Assignment8

| Build/Compile (Build Succeeded: 0 Error(s), 11 Warning(s))| **10** |
| --------------------- | ------ |
| Assignment compiles even though major pieces might be missing, or there are significant problems: **✔** |  **5** |
| The assignment contains only a few required elements or has significant problems. | **5**


| Assignment  | **<10 (missing lots) - 20 (minor)>** |
| --------------------- | ------ |
| 1. Implement the `ISampleData.CsvRows` property, loading the data from the `People.csv` file and returning each line as a single string. ✔ <br/>- Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔ <br/> - Using LINQ, skip the first row in the `People.csv`. ✔ <br/> - Be sure to appropriately handle resource (`IDisposable`) items correctly if applicable (and it may not be depending on how you implement it). ✔  | ✔ |
| 2. Implement `IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()` to return a **sorted**, **unique** list of states. ✔ <br/> Use `ISampleData.CsvRows` for your data source. ✔ <br/>- Don't forget the list should be unique. ✔ <br/>- Sort the list alphabetically ✔ <br/>- Include a test that leverages a hard coded list of Spokane based addresses. ✔ <br/>- Include a test that uses LINQ to verify the data is sorted correctly (do not use a hard coded list). ✔ | ✔ |
| 3. Implement `ISampleData.GetAggregateSortedListOfStatesUsingCsvRows()` to return a `string` that contains a **unique**, comma separated list of states. ✔ <br/>- Use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows` for your data source. ✔ <br/>- Given the array, consider using `string.Join` to combine the list into a single string. | ✔ |
| 4.Implement the `ISampleData.People` property to return all the items in `People.csv` as `Person` objects. ✔ <br/>- Use `ISampleData.CsvRows` as the source of the data. ✔ <br/>- Sort the list by State, City, Zip. (Sort the addresses first then select). ❌ <br/>- Be sure that `Person.Address` is also populated. <br/>- Adding null validation to all the `Person` and `Address` properties is **optional**. <br/>- Consider using `ISampleData.CsvRows` in your test to verify your results. | ✔ |
| 5. Implement `ISampleDate.FilterByEmailAddress(Predicate<string> filter)` to return a list of names where the email address matches the `filter`. ✔ <br/>- Use `ISampleData.People` for your data source. ✔ | ✔ |
| 6. Implement `ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)` to return a `string` that contains a **unique**, comma separated list of states. ✔ <br/>- Use the `people` parameter for your data source. ✔ <br/>- At a minimum, use `System.Linq.Enumerable.Aggregate` LINQ method to create your result. ❌ <br/>- Don't forget the list should be unique. ✔ <br/>- It is recommended that at a minimum you use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows` to validate your result. | ✔ |


| Project Analysis |   |
| -----------------| - |
| Disabled Warnings **** | **✔**
| Nullable Enabled: | **✔** |
| CodeReview Performed | **✔** |
| Appropriately handle all warnings | **❌** |
| Add the code analysis packages* |  **✔** |
*<sub><sup>`IntelliTectAnalyzer.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.CodeQuality.Analyzers.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.NetFramework.Analyzers.dll`. (It is okay to appropriately disable rules such as  localization/culture rules.)</sub></sup>


| ProjectName | Nullable | Solutions | Analyzers | Unexpected Disabled Warnings | Waning Level |
| -- | -- | -- | -- | -- | -- |
| Assignment | Enable | Assignment8.sln |  |  | 4 |
| Assignment.Tests | Enable | Assignment8.sln |  |  | 4 |


| Testing | **0** |
| --------------------- | ------ |
| Test Run Succeeded: Total tests=8, Passed=8, Failed=0, Skipped=0 | **98%** |
| Unit test project fails to compile, most test fails, or contains significant problems. | **10** |
| Most assigned work is tested or contain some errors. | **5** |
| All assigned work is tested | **5** |


| Extra Credit | **0** |
| --------------------- | ------ |
| Write awesome, well tested code that follows all the guidelines outlined in the book. | ❌ |
| Add support for an `IReadOnlyArray` interface (for you to define) that provides readonly access to the count and each element in the array.  Be sure to appropriately decorate it for covariance and contravariance. | ❌ |
| Implement all homework using async/await and multi-threading by defining a new `SampleDataAsync` class that implements `ISampleData`). Refactor both your SampleData and SampleDataAsync classes so there is minimal duplication.  This includes refactoring your tests so a significant amount of test code can be re-used.| ❌ |

**Pull request closed without merging**